moler/cmd/at/get_cell_id.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+"""
+AT+CREG?
+
+AT commands specification:
+google for: 3gpp specification 27.007
+(always check against latest version of standard)
+"""
+
+__author__ = 'Jakub Kochaniak'
+__copyright__ = 'Copyright (C) 2023, Nokia'
+__email__ = 'jakub.kochaniak@nokia.com'
+
+import re
+
+from moler.cmd.at.genericat import GenericAtCommand
+from moler.exceptions import ParsingDone
+
+
+class GetCellId(GenericAtCommand):
+    """
+    Command to get cell registration status. Example output:
+
+    +CREG: <n>,<stat>[,[<lac>],[<ci>],[<AcT>][,<cause_type>,<reject_cause>]]
+
+    OK
+    """
+    def __init__(self, connection=None, prompt=None, newline_chars=None, runner=None):
+        """Create instance of GetCellIdGsm class"""
+        super(GetCellId, self).__init__(connection, operation='execute', prompt=prompt,
+                                        newline_chars=newline_chars, runner=runner)
+        self.current_ret = dict()
+
+    def build_command_string(self):
+        return "AT+CREG?"
+
+    def on_new_line(self, line, is_full_line):
+        """
+        Method to parse command output. Will be called after line with command echo.
+
+        +CREG: <n>,<stat>[,[<lac>],[<ci>],[<AcT>][,<cause_type>,<reject_cause>]]
+
+        OK
+
+        Write your own implementation but don't forget to call on_new_line from base class
+
+        :param line: Line to parse, new lines are trimmed
+        :param is_full_line:  False for chunk of line; True on full line (NOTE: new line character removed)
+        :return: None
+        """
+        if is_full_line:
+            try:
+                self._parse_cell_data(line)
+            except ParsingDone:
+                pass
+        return super(GetCellId, self).on_new_line(line, is_full_line)
+
+    # <n>=3, where +CREG: <n>,<stat>,[<lac>],[<ci>],[<AcT>],<cause_type>,<reject_cause>
+    # +CREG: 3,5,,,,1,2
+    # +CREG: 3,5,"54DB","0F6B0578",15,1,2
+    _re_network_reg_mode_3 = \
+        re.compile(r'^\s*\+CREG:\s(?P<n>([0-9]+)),(?P<stat>([0-9]{1,2})),(\")?(?P<lac>([0-9A-Fa-f]*))?(\")?,'
+                   r'(\")?(?P<ci>([0-9A-Fa-f]*))?(\")?,(?P<AcT>([0-9]{1,2}))?,(?P<cause_type>([0-9]+)),'
+                   r'(?P<reject_cause>([0-9]+)).*$')
+
+    # <n>=2, where +CREG: <n>,<stat>,[<lac>],[<ci>],[<AcT>]
+    # +CREG: 2,5,,,
+    # +CREG: 2,5,"54DB","0F6B0578",7
+    _re_network_reg_mode_2 = \
+        re.compile(r'^\s*\+CREG:\s(?P<n>([0-9]+)),(?P<stat>([0-9]{1,2})),(\")?(?P<lac>([0-9A-Fa-f]*))?(\")?,'
+                   r'(\")?(?P<ci>([0-9A-Fa-f]*))?(\")?,(?P<AcT>([0-9]{1,2}))?.*$')
+
+    # <n>=0,1, where +CREG: <n>,<stat>
+    # +CREG: 1,1
+    _re_network_reg_mode_0_1 = re.compile(r'^\s*\+CREG:\s(?P<n>([0-9]+)),(?P<stat>([0-9]{1,2})).*$')
+
+    # +CREG: <output_data>
+    _re_cell_id_data = re.compile(r'^\s*\+CREG:\s(?P<output_data>\w+\S+).*$')
+
+    def _parse_cell_data(self, line):
```

Comment:
I would do it in this way
```
_re_creg = re.compile(r'\+CREG:\s(?P<CREG>.*)')


def _parse_cell_data(self, line):
    # +CREG: 1,1
    # +CREG: 2,5,,,
    # +CREG: 2,5,"54DB","0F6B0578",7
    # +CREG: 3,5,,,,1,2
    # +CREG: 3,5,"54DB","0F6B0578",15,1,2

    results_keys = ['n', 'stat', 'lac', 'ci', 'AcT']
    if self._regex_helper.match_compiled(self._re_results, line):
        self.current_ret = dict.fromkeys(results_keys, None)
        string_creg = self._regex_helper.groupdict()['CREG']
        self.current_ret['raw_output'] = string_creg
        creg = string_creg.split(',')
        for index, item in enumerate(creg, 0):
            if item:
                self.current_ret[results_keys[index]] = item
```

Reply:
Changed. I like this way, did not noticed that values in regex are always in the order, so "slitting" can make it right and faster. It is for sure better for code clearance (less lines).